client/allocrunner/taskrunner/template_hook.go

```diff
@@ -161,6 +161,25 @@ func (h *templateHook) newManager() (unblock chan struct{}, err error) {
 	return unblock, nil
 }
 
+// Exited runs when the task exits, but it may run again on restart.
+// Stop the template manager to prevent updates to templates of exited tasks.
+func (h *templateHook) Exited(ctx context.Context, req *interfaces.TaskExitedRequest, resp *interfaces.TaskExitedResponse) error {
+	h.managerLock.Lock()
+	defer h.managerLock.Unlock()
+
+	// Shutdown any created template
+	if h.templateManager != nil {
+		h.templateManager.Stop()
+
+		// Set templateManager to nil so it's recreated by Prestart on restart.
+		h.templateManager = nil
```

Comment:
This will block the task from restarting until the template has been *re-rendered.*

I'm a little worried this could cause correlated failures if something like this happens:

- A template dependency is unreachable
- Task dies for an unrelated reason

Prior to this change the task would restart after `restart.delay`, use whatever templated files it used last, and get back to work.

After this change the task would be blocked from restarting until all dependencies were reachable.

To put it another way: prior to this change client agents could handle _local_ "scheduling" decisions while the servers were in charge of _cluster_ scheduling decisions. (I think that's broadly true? Artifacts and Vault for example both allow disconnected local operation as long as they were able to successfully run _once._)

I _think_ we need to maintain that behavior although I'd love to be convinced otherwise! :sweat_smile: 

A quick peek at `TaskTemplateManager` makes me think adding the ability to `Pause/Resume` it would be a significant effort. Adding a new parameter or other bit of plumbing to make `TaskTemplateManager.Start()` blocking on `handleFirstRender` _optional_ might be an easier route.

Reply:
A potentially interesting case to consider:

1. Task A happily talks to Service X
2. Service X crashes or is otherwise partitioned from A
3. Task A crashes due to Service X going away
4. Task A is down for `restart.delay` duration
5. Service Y is registered as a replacement for Service X
6. Task A restarts and happily talks to Service Y

I think the behavior of Task A in this situation is:

<table>
    <tr>
        <td></td>
        <td>During restart.delay</td>
        <td>After restart.delay</td>
    </tr>
    <tr>
        <td>Before change</td>
        <td>Immediate restart</td>
        <td>Crash loop until Y is discovered</td>
    </tr>
    <tr>
        <td>After change</td>
        <td colspan=2>Blocked until Y is discovered</td>
    </tr>
</table>

I think in the happy cases of Service Y being ~immediately available, your changes are optimal!

However there are some convoluted situations in which I think it could make an outage worse:

If the only way to discover Service Y was being _rescheduled_ we're better off with the existing crash-loop behavior as that will eventually result in a rescheduling of the whole alloc. That's a pretty complex outage situation I think: Nomad would have fine connectivity but ...Task A's DC would lose connectivity to Consul and need rescheduling in a new DC? idk... in that complex of a failure scenario there may be a number of other failure conditions I'm not thinking of that alter the behavior of Task A.